applications/serial_lte_modem/src/slm_at_sms.c

```diff
@@ -0,0 +1,250 @@
+{
+	static uint16_t ref_number;
+	static uint8_t total_msgs;
+	static uint8_t count;
+	static char messages[MAX_CONCATENATED_MESSAGE - 1][SMS_MAX_PAYLOAD_LEN_CHARS];
+
+	ARG_UNUSED(context);
+
+	if (data == NULL) {
+		LOG_WRN("NULL data");
+		return;
+	}
+
+	if (data->type == SMS_TYPE_DELIVER) {
+		struct sms_deliver_header *header = &data->header.deliver;
+
+		if (!header->concatenated.present) {
+			sprintf(rsp_buf, "\r\n#XSMS: \"%02d-%02d-%02d %02d:%02d:%02d\",\"",
+				header->time.year, header->time.month, header->time.day,
+				header->time.hour, header->time.minute, header->time.second);
+			strcat(rsp_buf, header->originating_address.address_str);
+			strcat(rsp_buf, "\",\"");
+			strcat(rsp_buf, data->payload);
+			strcat(rsp_buf, "\"\r\n");
+			rsp_send(rsp_buf, strlen(rsp_buf));
+		} else {
+			LOG_DBG("concatenated message %d, %d, %d",
+				header->concatenated.ref_number,
+				total_msgs = header->concatenated.total_msgs,
+				header->concatenated.seq_number);
+			/* ref_number and total_msgs should remain unchanged */
+			if (ref_number == 0) {
+				ref_number = header->concatenated.ref_number;
+			}
+			if (ref_number != header->concatenated.ref_number) {
+				LOG_ERR("SMS concatenated message ref_number error: %d, %d",
+					ref_number, header->concatenated.ref_number);
+				goto done;
+			}
+			if (total_msgs == 0) {
+				total_msgs = header->concatenated.total_msgs;
+			}
+			if (total_msgs != header->concatenated.total_msgs) {
+				LOG_ERR("SMS concatenated message total_msgs error: %d, %d",
+					total_msgs, header->concatenated.total_msgs);
+				goto done;
+			}
+			if (total_msgs > MAX_CONCATENATED_MESSAGE) {
+				LOG_ERR("SMS concatenated message no memory: %d", total_msgs);
+				goto done;
+			}
+			/* seq_number should start with 1 but could arrive in random order */
+			if (header->concatenated.seq_number == 0 ||
+			    header->concatenated.seq_number > total_msgs) {
+				LOG_ERR("SMS concatenated message seq_number error: %d, %d",
+					header->concatenated.seq_number, total_msgs);
+				goto done;
+			}
+			if (header->concatenated.seq_number == 1) {
+				sprintf(rsp_buf, "\r\n#XSMS: \"%02d-%02d-%02d %02d:%02d:%02d\",\"",
+					header->time.year, header->time.month, header->time.day,
+					header->time.hour, header->time.minute,
+					header->time.second);
+				strcat(rsp_buf, header->originating_address.address_str);
+				strcat(rsp_buf, "\",\"");
+				strcat(rsp_buf, data->payload);
+				count++;
+			} else {
+				strcpy(messages[header->concatenated.seq_number - 2],
+					data->payload);
+				count++;
+			}
+			if (count == total_msgs) {
+				for (int i = 0; i < (total_msgs - 1); i++) {
+					strcat(rsp_buf, messages[i]);
+				}
+				strcat(rsp_buf, "\"\r\n");
+				rsp_send(rsp_buf, strlen(rsp_buf));
+			} else {
+				return;
+			}
+done:
+			ref_number = 0;
+			total_msgs = 0;
+			count = 0;
+		}
+	} else if (data->type == SMS_TYPE_STATUS_REPORT) {
+		LOG_INF("STATUS_REPORT receievd");
```

Comment:
`receievd` -> `received`

I would also rather say `Status report received` but I'll leave that decision to you

Reply:
Thanks, I'll fix both together with other comments